alacritty_terminal/src/tty/windows/child.rs

```diff
@@ -0,0 +1,180 @@
+        let success = unsafe {
+            RegisterWaitForSingleObject(
+                &mut wait_handle,
+                child_handle,
+                Some(child_exit_callback::<F>),
+                Box::into_raw(wait_notify) as PVOID,
+                INFINITE,
+                WT_EXECUTEINWAITTHREAD | WT_EXECUTEONLYONCE,
+            )
+        };
+
+        if 0 == success {
+            Err(Error::last_os_error())
+        } else {
+            Ok(HandleWaitSignal { wait_handle: AtomicPtr::from(wait_handle) })
+        }
+    }
+}
+
+impl Drop for HandleWaitSignal {
+    fn drop(&mut self) {
+        unsafe {
+            // Cancel WinAPI wait for child process exiting
+            UnregisterWait(self.wait_handle.load(Ordering::Relaxed));
+        }
+    }
+}
+
+pub(crate) struct ChildProcessState {
+    _on_exit: HandleWaitSignal,
+    events: Receiver<ChildEvent>,
+}
+
+impl ChildProcessState {
+    pub fn new(subprocess_handle: HANDLE) -> Result<ChildProcessState, Error> {
+        let (sender, receiver) = channel();
+        let on_exit = HandleWaitSignal::new(subprocess_handle, move || {
+            if let Err(e) = sender.send(ChildEvent::Exited) {
+                warn!(
+                    "An error occurred while attempting to notify about child process z
+                    termination: {}",
+                    e
+                );
+            }
+        })?;
+
+        Ok(ChildProcessState { _on_exit: on_exit, events: receiver })
+    }
+
+    pub fn events(&self) -> &Receiver<ChildEvent> {
+        &self.events
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::io::Error;
+    use std::ptr;
+    use std::time::Duration;
+
+    use std::sync::mpsc::channel;
+
+    use widestring::U16CString;
+
+    use winapi::shared::ntdef::LPWSTR;
+    use winapi::um::processthreadsapi::{
+        CreateProcessW, TerminateProcess, PROCESS_INFORMATION, STARTUPINFOW,
+    };
+
+    use super::*;
+
+    fn make_cmd_process() -> Result<HANDLE, Error> {
+        let mut pi: PROCESS_INFORMATION = Default::default();
+        let mut si: STARTUPINFOW = Default::default();
+        let cmdline = U16CString::from_str("cmd.exe").unwrap();
+
+        unsafe {
+            if 0 == CreateProcessW(
+                ptr::null(),
+                cmdline.as_ptr() as LPWSTR,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                false as i32,
+                Default::default(),
+                ptr::null_mut(),
+                ptr::null_mut(),
+                &mut si as *mut STARTUPINFOW,
+                &mut pi as *mut PROCESS_INFORMATION,
+            ) {
+                Err(Error::last_os_error())
+            } else {
+                Ok(pi.hProcess)
+            }
+        }
+    }
+
+    // TODO: This is more of an integration test since it spawns new 'cmd.exe'
+    //       process and heavily uses Win32 API. Should this be ignored
+    //       by default? Worst-case, this will timeout after `WAIT_TIMEOUT`.
```

Comment:
I'd assume just the shell part of `cmd.exe` is used? Or does this actually pop open a new window?

And is the runtime of this test significant? Is the `WAIT_TIMEOUT` just a maximum and it usually runs in under 200ms? If so, why do we choose 200ms as timeout here?

Reply:
This test could run any windows process as long as it can be spawned and does not immediately exit. I use `cmd.exe` because it is available on the path easily on any Windows machine. If something goes wrong with the test run when process gets spawned but not killed, it will be running in the background until the test runner exits.

The timeout was just picked in an arbitrary fashion. I know that by default `cargo test` times tests out after a minute, but there is no point waiting for that long. The successful path through the test takes just a few milliseconds so I just scaled it by ~2 orders of magnitudes, which is still way faster failure than 1 minute.

Reply:
Makes sense, thanks for the details. I think if this is just spawning a subprocess that is completely silent (just the way Alacritty silently spawns 'cmd.exe' without anything else opening), it should be fine.

It's true that it is more like an integration test, but I don't think there's much wrong with having a quick running integration test at this place. The problem with integration is usually just runtime and if what you're saying is correct, this doesn't really impact runtime of the tests.

Reply:
I'm perfectly fine mixing in integration tests here, especially since it's a `windows` module. However we could also start collecting integration style tests in a `tests` directory as recommended by the [Rust Lang Book](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests), though again, I see no need for this here.

Reply:
> However we could also start collecting integration style tests in a tests directory as recommended by the Rust Lang Book, though again, I see no need for this here.

We already do that. However I'm not sure if that's necessary here.